app-v2/components/CodeMirror/index.tsx

```diff
@@ -0,0 +1,195 @@
+import 'codemirror/keymap/sublime';
+import 'codemirror/lib/codemirror.css';
+import 'codemirror/mode/meta';
+import 'codemirror/theme/monokai.css';
+import React from 'react';
+
+import { ban, keyWords, maxLineNum, operators } from '@appv2/config/nebulaQL';
+
+import './index.less';
+
+interface IProps {
+  options?: object;
+  value: string;
+  ref?: any;
+  width?: string;
+  height?: string;
+  onShiftEnter?: () => void;
+  onChange?: (value: string) => void;
+  onBlur?: (value: string) => void;
+  onChangeLine?: () => void;
+}
+
+export default class ReactCodeMirror extends React.PureComponent<IProps, any> {
+  codemirror;
+  editor;
+  textarea;
+  constructor(props) {
+    super(props);
+  }
+  public componentDidMount() {
+    CodeMirror.defineMode('nebula', () => {
+      return {
+        token: stream => {
+          if (stream.eatSpace()) {
+            return null;
+          }
+          stream.eatWhile(/[\$:\w\u4e00-\u9fa5]/);
+          const cur = stream.current();
+          if (keyWords.some(item => item === cur)) {
+            return 'keyword';
+          } else if (operators.some(item => item === cur)) {
+            return 'def';
+          } else if (ban.some(item => item === cur)) {
+            return 'error';
+          }
+          stream.next();
+        },
+      };
+    });
+
+    CodeMirror.registerHelper('hint', 'nebula', cm => {
+      const cur = cm.getCursor();
+      const token = cm.getTokenAt(cur);
+      const str = token.string;
+      const start = token.start;
+      const end = cur.ch;
+
+      if (str === '') {
+        return;
+      }
+
+      const list = [...keyWords, ...operators, ...ban].filter(item => {
+        return item.indexOf(str) === 0;
+      });
+
+      if (list.length) {
+        return {
+          list,
+          from: CodeMirror.Pos(cur.line, start),
+          to: CodeMirror.Pos(cur.line, end),
+        };
+      }
+    });
+    this.renderCodeMirror();
+  }
+  renderCodeMirror() {
+    // parameters of the combined
+    const options = {
+      tabSize: 2,
+      fontSize: '14px',
+      autoCloseBrackets: true,
+      matchBrackets: true,
+      showCursorWhenSelecting: true,
+      lineWrapping: true,
+      // show number of rows
+      lineNumbers: true,
+      fullScreen: true,
+      mode: 'nebula',
+      ...this.props.options,
+    };
+    this.editor = CodeMirror.fromTextArea(this.textarea, options);
+    // Getting CodeMirror is used to get some of these constants
+    this.codemirror = CodeMirror;
+    // event
+    this.editor.on('change', this.codemirrorValueChange);
+    this.editor.on('keydown', this.keydown);
+    this.editor.on('blur', this.blur);
+    const { value, width, height } = this.props;
+    this.editor.setValue(value || '');
+    if (width || height) {
+      // set size
+      this.editor.setSize(width, height);
+    }
+  }
+
+  blur = instance => {
+    if (this.props.onBlur) {
+      this.props.onBlur(instance.doc.getValue());
+    }
+  };
```

Comment:
it's more elegant to use optional chaining instead of `a && a.b && a.b()`
```ts
blur = instance => this.props.onBlur?.(instance.doc.getValue());
```